### DISCUSSION TITLE
Enabling GADP Agent standalone execution?

### DISCUSSION THREAD

**Discussion by hedgeberg** [NationalSecurityAgency/ghidra#3076] [Debugger]
What would it take to make it possible to launch one of the provided GADP agents without launching the full Ghidra installation? Looking at the source for GDB's GADP agent, it looks like it was intended to be launched from the command line (see [here](https://github.com/NationalSecurityAgency/ghidra/blob/e7488245fd3e85dea6050e0dd66bb4a4dbeeb53b/Ghidra/Debug/Debugger-agent-gdb/src/main/java/agent/gdb/gadp/GdbGadpServer.java#L123)) but I can't find any documentation or code indicating that the standalone operation feature survived into release. Being able to do something like launch one of the GADP agents via something like `"${SCRIPT_DIR}"/support/launch.sh bg Ghidra "${MAXMEM}" "" ghidra.GADPGDB"$@"` would be really helpful for setups that involve complex networking where the intent is to debug something running on a target device of a different architecture. 

Also, if there were an option to launch the GADP agents without producing the server popup, that would make life a lot easier for some of the stranger debugging setups I've created.

Is there any plan currently to add standalone launching of the agents, or "headless" launching, to the 10.0 release? If not, what would it take to insert a new launcher option to the set of supported options?

**Comment by d-millar** [comment]
@hedgeberg I'm not sure this is exactly what you're looking for, but, if you build the agent as a jar, you can run it standalone using "java -jar nodepJar".  The build command is:
    gradle nodepJar
or, if you prefer to restrict it to a single agent:
    gradle Debugger-agent-gdb:nodepJar (or pick the name of the agent you want)

**Reply by hedgeberg** [reply]
Thanks, and noted. This is good to know, but it would probably be a good idea to either document this somewhere, and/or to add a step for building the agents as no-dependency JARs to the default buildGhidra gradle routine. My ideal scenario was for both a standalone option like you described to be present, but I wasn't sure how much work it would be to make that feasible. I will say that I still like the idea of being able to launch GADP agent instances from the ghidra installation without launching everything else, but a standalone option fixes 95% of my concerns here.

**Reply by d-millar** [reply]
Agreed re documentation - I had to search around to remember how to do it myself.  In your second scenario, I'm not quite picturing the exact workflow.  Can you give me a little more detail?

**Reply by astrelsky** [reply]
I tried to run `gradle :Debugger-agent-gdb:executableJar` but it ended up failing at least on gradle 7.0.2

```
> Task :Debugger-agent-gdb:nodepJar FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':Debugger-agent-gdb:nodepJar'.
> Entry images/checkmark_green.gif is a duplicate but no duplicate handling strategy has been set. Please refer to https://docs.gradle.org/7.0.2/dsl/org.gradle.api.tasks.Copy.html#org.gradle.api.tasks.Copy:duplicatesStrategy for details.
```

**Reply by hedgeberg** [reply]
Yeah, I did something very cursed to get some simulacrum of what I was looking for a couple weeks back -- I wrote a bash script that configures the classpath and just launches the GADP agent using that. It looks sorta like this:

```bash
#!/usr/bin/bash                                                                                                                                                                  #CMD="/home/hedge/projectdir/custom_gdb_build"
#EXTOPS="--nx"
HOST=localhost
PORT=15432
PREFIX=$GHIDRA_INSTALL_DIR/
CLASSPATH="${PREFIX}Ghidra/Framework/Utility/lib/Utility.jar"
CLASSPATH="${CLASSPATH}:${PREFIX}Ghidra/Framework/Docking/lib/Docking.jar"
CLASSPATH="${CLASSPATH}:${PREFIX}Ghidra/Framework/Docking/lib/timingframework-1.0.jar"
#<SNIP>...
CLASSPATH="${CLASSPATH}:${PREFIX}Ghidra/Debug/ProposedUtils/lib/ProposedUtils.jar"                                                                                                

java -cp "${CLASSPATH}" agent.gdb.gadp.GdbGadpServer --gadp-args -H "${HOST}" -p "${PORT}"  -x 
```

(Don't judge me, I'm very bad at Java Things(tm) and just wanted to get this stupid setup of mine working)

Which let me launch an instance of GADPGDB from the command line at runtime. 

My workflow looked sorta like:
-Realize I need to figure something out on this binary using the debugger
-Close project and open it in the Debugger tool
-Launch gdbserver on the target device
-Launch GADP agent in a VM attached to the device from the command line
-Launch a custom build of GDB for this device, then attach it to the agent via `new-ui mi2`
-Connect to the GADP agent from outside the VM using GADP/TCP

After a while I got a laptop with better specs, so I started running Ghidra inside the VM, but I still found it convenient to be able to launch the GADP server from outside of the JVM instance Ghidra was in, and then connecting to that over GADP/TCP. It's been a lot more stable than trying to run gdb from inside the JVM or letting ghidra spin up the GADP agent instance, for some reason. Also, Ghidra only shows which PTY to attach to in new-ui mode in the project browser's log, which adds a frustrating step when trying to initialize my debugging chain.

**Reply by hedgeberg** [reply]
@astrelsky tried deleting the .gif in question?

**Reply by hedgeberg** [reply]
Attempting the same thing, I get a different error: 
```
> Task :Debugger-agent-gdb:configureNodepJar FAILED

FAILURE: Build failed with an exception.

* Where:
Build file '/home/hedge/Documents/ghidra_src/Ghidra/Debug/Debugger-agent-gdb/build.gradle' line: 55

* What went wrong:
Execution failed for task ':Debugger-agent-gdb:configureNodepJar'.
> Resolving dependency configuration 'runtimeOnly' is not allowed as it is defined as 'canBeResolved=false'.
  Instead, a resolvable ('canBeResolved=true') dependency configuration that extends 'runtimeOnly' should be resolved.
```

**Reply by d-millar** [reply]
Hmmm, the move to gradle 7 may have thrown a wrench in some of our less-used build scripts.  @nsadeveloper789 is out for a few days, but will have him look at it when he gets back.  (He's definitely the gradle guy on the debugger front.)

@hedgeberg That description helps - give us a couple of days, see if we can work something out.

**Reply by nsadeveloper789** [reply]
@hedgeberg In your case, I think you just need to pull, we fixed that issue in a commit labeled GP-960.
@astrelsky I haven't seen this one before. I'm not sure why, maybe Gradle changed some defaults in later versions on us. According to the docs and some online examples, it looks like you just need to add `duplicatesStrategy = 'warn'` into the nodepJar tasks of each respective build.gradle file. If I manage to replicate the issue here, I can try it myself, but I'm a bit busy having just returned....

**Reply by hedgeberg** [reply]
@nsadeveloper789 You're right. Cleaned, pulled, and rebuilt, and now I'm getting the same issue that @astrelsky was. I'll experiment with adding the necessary change on my end and see if I can get it working.

**Reply by hedgeberg** [reply]
Build appears to have succeeded after adding `duplicatesStrategy = 'warn'` to the nodepJar task in `Ghidra/Debug/Debugger-agent-gdb/build.gradle`, but it gives a lot of warnings of the form: 

`Encountered duplicate path "images/checkmark_green.gif" during copy operation configured with DuplicateStrategy.WARN` 

Is there a "correct" way to handle this kind of duplicate conflict within Gradle? Because even if this solution works, it doesn't seem like its the way Gradle "wants" us to be doing it, which makes me worried that upgrading in the future will cause further problems.

**Reply by nsadeveloper789** [reply]
Uh, "correctness" is in the eye of the beholder? Honestly, I'm not sure. At the moment, I was just trying to find the quick fix for you. Unfortunately, the dependency graph among our projects kind of sucks, so the no-dep jar winds up including a whole lot more stuff than it actually needs, and the .gif in question is one of them. What I'll probably wind up doing is just explicitly excluding certain paths, e.g. `images/*` that I know aren't needed, and that are likely to produce duplicates. If I run into duplicate .class files, then it would seem something else is seriously wrong.

**Reply by hedgeberg** [reply]
So... on the one hand, it seems the standalone jar runs just fine...

On the other hand, uh, I hate to do this to you guys, but here's some other snippets from that warning log..... 

```
Encountered duplicate path "org/osgi/service/url/URLConstants.class" during copy operation configured with DuplicatesStrategy.WARN
Encountered duplicate path "org/osgi/service/url/URLStreamHandlerService.class" during copy operation configured with DuplicatesStrategy.WARN
Encountered duplicate path "org/osgi/service/url/URLStreamHandlerSetter.class" during copy operation configured with DuplicatesStrategy.WARN
Encountered duplicate path "org/osgi/service/url/package-info.class" during copy operation configured with DuplicatesStrategy.WARN
Encountered duplicate path "org/osgi/util/tracker/AbstractTracked.class" during copy operation configured with DuplicatesStrategy.WARN
Encountered duplicate path "org/osgi/util/tracker/BundleTracker$Tracked.class" during copy operation configured with DuplicatesStrategy.WARN
Encountered duplicate path "org/osgi/util/tracker/BundleTracker.class" during copy operation configured with DuplicatesStrategy.WARN
Encountered duplicate path "org/osgi/util/tracker/BundleTrackerCustomizer.class" during copy operation configured with DuplicatesStrategy.WARN
Encountered duplicate path "org/osgi/util/tracker/ServiceTracker$AllTracked.class" during copy operation configured with DuplicatesStrategy.WARN
Encountered duplicate path "org/osgi/util/tracker/ServiceTracker$Tracked.class" during copy operation configured with DuplicatesStrategy.WARN
Encountered duplicate path "org/osgi/util/tracker/ServiceTracker.class" during copy operation configured with DuplicatesStrategy.WARN
Encountered duplicate path "org/osgi/util/tracker/ServiceTrackerCustomizer.class" during copy operation configured with DuplicatesStrategy.WARN
Encountered duplicate path "org/osgi/util/tracker/package-info.class" during copy operation configured with DuplicatesStrategy.WARN
Encountered duplicate path "org/osgi/util/tracker/packageinfo" during copy operation configured with DuplicatesStrategy.WARN
```

Sorry @nsadeveloper789 but it seems like there's a bunch of re-included .class files from org/osgi and OSGI-OPT/src/org/osgi. Anything I can do to help triage?

**Reply by hedgeberg** [reply]
Went ahead and dumped the whole debug log, commands to generate this log (from the ghidra root dir) were:

```bash
rm -rf ./Ghidra/Debug/Debugger-agent-gdb/build
gradle :Debugger-agent-gdb:executableJar | tee -a ./agent_build_log.txt
```

Log file is linked below, let me know if I can get you anything else.

[agent_build_log.txt](https://github.com/NationalSecurityAgency/ghidra/files/6587111/agent_build_log.txt)

**Reply by nsadeveloper789** [reply]
eek. OK. Thanks for the heads-up. It should work fine for you, but for the sake of cleanliness, that's something I should look at more carefully.
